### ISSUE TITLE
How to return a literal value as a constrained type?

### ISSUE THREAD

**Issue by dycw** [microsoft/pyright#2010]
Hi all,

```python
from __future__ import annotations

from typing import TypeVar


T = TypeVar("T", int, str)


def foo(x: T) -> T:
    if isinstance(x, int):
        return 2 * x
    else:
        return x + "!"


def bar(x: T) -> T:
    if isinstance(x, int):
        return 2 * x
    else:
        return "just-a-str" # Expression type "Literal['just-a-str']" cannot be assigned to return type...
```

I would like to know how to nicely return `"just-a-str"` in `bar()`, or is this an anti-pattern? I know the usual hacks of `# type-ignore` or `cast(T, ...)`, but would prefer not to.

Version = 1.1.150
OS = Ubuntu

**Comment by erictraut** [comment]
The behavior you're seeing is a side effect of the way pyright handles constrained type variables. Conditional type variables are very challenging to handle within a single-pass (lazy) type checker that is designed to be used as the basis for a language server. Mypy is a multi-pass type checker, and it handles constrained type variables by performing multiple full passes over a function if it involves a constrained type variable. Mypy's approach has some advantages, but it also has some major downsides. It's much slower. It can produce multiple redundant errors (once for each constraint in the TypeVar). And it also doesn't allow you to point to a given symbol in the function and ask "what type is this" because it differs from one pass to the next.

Pyright solves the challenge of constrained type variables by tracking "conditional types" that are conditioned on specific constraints. In the sample above, `x` is type `T` initially. It is narrowed to type `int*` based on the `isinstance` check. The `*` means that it's conditional — it's an `int` only if the original `x` used the `int` constraint of `T`. This condition is propagated to the type of expression `2 * x`, so its type is also `int*`.

There is no condition associated with `"just-a-str"` because the type of this expression isn't derived from `x` or any other variable of type `T`.

I don't see an elegant way to work around this limitation. Nor do I see a feasible way to address this limitation in pyright. 

As you noted, you can use a `cast(T, ...)`.

```python
    return cast(T, "hi")
```

Here are a few other ideas that are perhaps a bit less ugly but still not great.

```python
    return type(x)("hi")
```

or

```python
    return type(x)() + "hi"
```

or

```python
    return cast(type(x), "hi")
```

**Comment by dycw** [comment]
Thanks as usual @erictraut.
